kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala

```diff
@@ -244,8 +244,21 @@ class BatchJobSubmission(
       }
 
       try {
-        killMessage = killBatchApplication()
         builder.close()
+      } catch {
+        case e: Exception => warn("failed to close engine process builder", e)
+      }
+      MetricsSystem.tracing(_.decCount(
+        MetricRegistry.name(OPERATION_OPEN, statement.toLowerCase(Locale.getDefault))))
+
+      // fast fail
+      if (isTerminalState(state)) {
+        killMessage = (false, s"batch $batchId is already terminal so can not kill it.")
+        return
```

Comment:
seems that we can not close builder before `killBatchApplication`.

For yarn-cluster, if we close builder, then the async thread will terminate because the builder process is not alive.

And then the state is terminalState, but the application is still running.

Reply:
good catch I considered before but missed it last commit ..